### ISSUE TITLE
Displaying Exception page causes an InvalidOperationException when Kestrel AllowSynchronousIO is set to false

### ISSUE THREAD

**Issue by TheBlueSky** [aspnet/AspNetCore#8379]
### Describe the bug
I have an ASP.NET Core 2.2 application in which I set `AllowSynchronousIO` to `false` in Kestrel options. Now, whenever the application has an error and tries to display the `DeveloperExceptionPage`, I get the following error:
```
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[3]
      An exception was thrown attempting to display the error page.
System.InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.
```
### To Reproduce
Steps to reproduce the behaviour:
1. Create a new ASP.NET Core 2.2 Web application; e.g., using Visual Studio File > New > Project
2. Change the code in `Program` class to be
```csharp
public static async Task Main(string[] args)
{
    await CreateWebHostBuilder(args).Build().RunAsync();
}

public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
    WebHost.CreateDefaultBuilder(args)
        .ConfigureKestrel(options => { options.AllowSynchronousIO = false; })
        .UseStartup<Startup>();
```
3. Add `<LangVersion>latest</LangVersion>` inside a `PropertyGroup`
4. Create a `Person` class or whatever you would like to call it; keep it empty, we won't use its members, but only to inject it without registeration to get some Exception
5. Add the following constructor to `IndexModel` class in `Pages/Index.cshtml.cs` file:
```csharp
public IndexModel(Person person)
{
}
```
6. Run the application with `dotnet run`, to see the log message in the Console
7. You'll see 2 Exceptions; the first of which is the error I'm reporting

### Expected behaviour
See the `DeveloperExceptionPage`

### Additional context
SDK version 2.2.104
Runtime version 2.2.2

### Sample application to reproduce the issue
[WebApplication3.zip](https://github.com/aspnet/AspNetCore/files/2949302/WebApplication3.zip)

**Comment by davidfowl** [comment]
Thanks for the report
- There are a bunch of known issues with `AllowSynchronousIO` in 2.2 that are being resolved in 3.0 and it's likely you're running into that.
- What are the actual exceptions? (can you paste the full stack trace in here)

**Comment by TheBlueSky** [comment]
Here is the whole log which includes both the exception; the one due to failed activation (the expected one) and the one from the `DeveloperExceptionPage`
```
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET https://localhost:5001/
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'Page: /Index'
info: Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker[3]
      Route matched with {page = "/Index"}. Executing page /Index
info: Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker[4]
      Executed page /Index in 7.2566ms
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'Page: /Index'
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
System.InvalidOperationException: Unable to resolve service for type 'WebApplication3.Person' while attempting to activate 'WebApplication3.Pages.IndexModel'.
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(IServiceProvider sp, Type type, Type requiredBy, Boolean isDefaultParameterRequired)
   at lambda_method(Closure , IServiceProvider , Object[] )
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.DefaultPageModelActivatorProvider.<>c__DisplayClass1_0.<CreateActivator>b__0(PageContext context)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.DefaultPageModelFactoryProvider.<>c__DisplayClass3_0.<CreateModelFactory>b__0(PageContext pageContext)
   at Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.CreateInstance()
   at Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[3]
      An exception was thrown attempting to display the error page.
System.InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder)
   at System.IO.StreamWriter.Write(String value)
   at Microsoft.AspNetCore.Diagnostics.RazorViews.ErrorPage.ExecuteAsync()
   at Microsoft.Extensions.RazorViews.BaseView.ExecuteAsync(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HLL5C735JO0O", Request id "0HLL5C735JO0O:00000001": An unhandled exception was thrown by the application.
System.InvalidOperationException: Unable to resolve service for type 'WebApplication3.Person' while attempting to activate 'WebApplication3.Pages.IndexModel'.
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(IServiceProvider sp, Type type, Type requiredBy, Boolean isDefaultParameterRequired)
   at lambda_method(Closure , IServiceProvider , Object[] )
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.DefaultPageModelActivatorProvider.<>c__DisplayClass1_0.<CreateActivator>b__0(PageContext context)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.DefaultPageModelFactoryProvider.<>c__DisplayClass3_0.<CreateModelFactory>b__0(PageContext pageContext)
   at Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.CreateInstance()
   at Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.RazorPages.Internal.PageActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 227.4091ms 500
```

**Comment by davidfowl** [comment]
Yep, these issues are fixed in 3.0.

**Comment by TheBlueSky** [comment]
Thanks David.

Is this something that can be ported back to a patch release to 2.2?

**Comment by davidfowl** [comment]
No, the problems are extremely broad and required new features in the framework. This is why we pretty much disabled it by default in 2.2. This specific exception is just the tip of the iceberg.

cc @Tratcher have you encountered this on 3.0?

**Comment by Tratcher** [comment]
Yes, I fixed it by buffering the error page.

**Comment by Tratcher** [comment]
Closing as fixed in 3.0 preview2
